lib/web/automaticupgrades.go

```diff
@@ -0,0 +1,144 @@
+// Copyright 2023 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package web
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/julienschmidt/httprouter"
+	"golang.org/x/mod/semver"
+
+	"github.com/gravitational/teleport"
+	"github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/constants"
+	versionlib "github.com/gravitational/teleport/integrations/kube-agent-updater/pkg/version"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+)
+
+const defaultChannelTimeout = 5 * time.Second
+
+// automaticUpgrades implements a version server in the Teleport Proxy.
+// It is configured through the Teleport Proxy configuration and tells agent updaters
+// which version they should install.
+func (h *Handler) automaticUpgrades(w http.ResponseWriter, r *http.Request, p httprouter.Params) (interface{}, error) {
+	if h.cfg.AutomaticUpgradesChannels == nil {
+		return nil, trace.AccessDenied("This proxy is not configured to server automatic upgrades channels.")
+	}
+
+	// The request format is "<channel name>/{version,critical}"
+	// As <channel name> might contain "/" we have to split, pop the last part
+	// and re-construct the channel name.
+	channelAndType := p.ByName("request")
+
+	reqParts := strings.Split(strings.Trim(channelAndType, "/"), "/")
+	if len(reqParts) < 2 {
+		return nil, trace.BadParameter("path format should be /webapi/automaticupgrades/<channel>/{version,critical}")
+	}
+	requestType := reqParts[len(reqParts)-1]
+	channelName := strings.Join(reqParts[:len(reqParts)-1], "/")
```

Comment:
We can't change this without breaking things for customers, can we?
It feels a bit confusing to do all this splits, joins and sub-slices
We could use query parameters to simplify
Eg, `/webapi/automaticaupgrades/version?channel=stable/cloud`
Or the other way around
Eg `/webapi/automaticaupgrades/stable/cloud?critical`

Can we extract this to a function and unit test it?

Reply:
> We can't change this without breaking things for customers, can we?

Yeah, the protocol is kinda dumb and we need to stick to the original design. It was made under the (false) assumption that no logic was needed and we could just hit the s3 bucket. If I were to design from scratch I'd chose one of the path you're suggesting